configuration/postgres/postgres.go

```diff
@@ -0,0 +1,415 @@
+		return "", fmt.Errorf("cannot subscribe to empty trigger")
+	}
+
+	var subscribeID string
+	p.configLock.Lock()
+	for _, trigger := range triggers {
+		notificationChannel := "listen " + trigger
+		if sub, isActive := p.isSubscriptionActive(req); isActive {
+			if oldStopChan, ok := p.subscribeStopChanMap[sub]; ok {
+				close(oldStopChan.(chan struct{}))
+			}
+		}
+		stop := make(chan struct{})
+		subscribeID = uuid.New().String()
+		p.subscribeStopChanMap[subscribeID] = stop
+		p.ActiveSubscriptions[trigger] = &subscription{
+			uuid:    subscribeID,
+			trigger: trigger,
+			keys:    req.Keys,
+		}
+		go p.doSubscribe(ctx, req, handler, notificationChannel, trigger, subscribeID, stop)
+	}
+	p.configLock.Unlock()
+	return subscribeID, nil
+}
+
+func (p *ConfigurationStore) Unsubscribe(ctx context.Context, req *configuration.UnsubscribeRequest) error {
+	p.configLock.Lock()
+	defer p.configLock.Unlock()
+	if oldStopChan, ok := p.subscribeStopChanMap[req.ID]; ok {
+		delete(p.subscribeStopChanMap, req.ID)
+		close(oldStopChan.(chan struct{}))
+		for k, v := range p.ActiveSubscriptions {
+			if v.uuid == req.ID {
+				channel := "unlisten " + v.trigger
+				conn, err := p.client.Acquire(ctx)
+				if err != nil {
+					return fmt.Errorf("Error acquiring connection: %v", err)
+				}
+				defer conn.Release()
+				_, err = conn.Exec(ctx, channel)
+				if err != nil {
+					return fmt.Errorf("Error unlistening to channel: %v", err)
+				}
+				delete(p.ActiveSubscriptions, k)
+			}
+		}
+		return nil
+	}
+	return fmt.Errorf("unable to find subscription with ID : %v", req.ID)
+}
+
+func (p *ConfigurationStore) doSubscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler, channel string, trigger string, subscription string, stop chan struct{}) {
+	conn, err := p.client.Acquire(ctx)
+	if err != nil {
+		p.logger.Errorf("Error acquiring connection:", err)
+		return
+	}
+	defer conn.Release()
+	_, err = conn.Exec(ctx, channel)
+	if err != nil {
+		p.logger.Errorf("Error listening to channel:", err)
+		return
+	}
+
+	for {
+		notification, err := conn.Conn().WaitForNotification(ctx)
+		if err != nil {
+			if !(pgconn.Timeout(err) || errors.Is(ctx.Err(), context.Canceled)) {
+				p.logger.Errorf("Error waiting for notification:", err)
+			}
+			return
+		}
+		p.handleSubscribedChange(ctx, handler, notification, trigger, subscription)
+	}
+}
+
+func (p *ConfigurationStore) handleSubscribedChange(ctx context.Context, handler configuration.UpdateHandler, msg *pgconn.Notification, trigger string, subscription string) {
+	defer func() {
+		if err := recover(); err != nil {
+			p.logger.Errorf("panic in handleSubscribedChange method and recovered: %s", err)
+		}
+	}()
+	payload := make(map[string]interface{})
+	err := json.Unmarshal([]byte(msg.Payload), &payload)
+	if err != nil {
+		p.logger.Errorf("Error in UnMarshal: ", err)
+		return
+	}
+
+	var key, value, version string
+	m := make(map[string]string)
```

Comment:
Is there a possibility that multiple values are returned for the same key and if so should we arrange by version descending and return latest?

Reply:
Yes. The challenge is version is a string. Issue https://github.com/dapr/dapr/issues/5183#issuecomment-1251328400 is created to mandate version to follow semantic versioning so that version can be compared. As it cannot be reliable compared today, documentation is updated to ensure single value of key to be maintained in the configuration store.

Reply:
@mukundansundar  - marked this as resolved by mistake. please comment if this needs more attention